File: problems/sliding_window/minSubArrayLen.ts
/*
 *
 * 7 , [2,3,1,2,4,3]
 * 4 , [1,4,4]
 * 11 , [1,1,1,1,1,1,1,1]
 * 10 , [1,2,3,4]
 * 5 , [1,2,1,3]
 *
 */

export function minSubArraylen(target: number, nums: number[]) {
  let min = Infinity;

  for (let i = 0; i < nums.length; i++) {
    // current number is equal or bigger than the target find min value
    if (nums[i] >= target) return 1;

    let windowSize = 0;
    let sum = 0;

    while (sum < target && i + windowSize < nums.length) {
      sum += nums[i + windowSize];
      windowSize++;
    }

    if (min > windowSize && sum >= target) {
      min = windowSize;
    }
  }

  if (min === Infinity) min = 0;

  return min;
}

// better solution
export function minSubArraylenSolution(target: number, nums: number[]) {
  let windowSize = Infinity;

  let start = 0;
  let totalSum = 0;

  for (let end = 0; end < nums.length; end++) {
    totalSum += nums[end];

    while (totalSum >= target) {
      // get current window size
      let size = end + 1 - start;

      windowSize = Math.min(size, windowSize);

      // move the slide window
      totalSum -= nums[start];
      start++;
    }
  }

  if (windowSize === Infinity) return 0;

  return windowSize;
}

File: problems/LeetCode/sliding_window/minSubArrayLen.py


from cmath import inf
from typing import List


def minSubArrayLen(self, target: int, nums: List[int]) -> int:
		sum = 0
		window_size = inf
		window_start = 0

		for window_end in range(len(nums)):
				
				sum += nums[window_end]

				while sum - nums[window_start] >= target:
						sum -= nums[window_start]
						window_start += 1
				
				if sum >= target:
						window_size = min(window_size, window_end - window_start + 1)

		return window_size if sum >= target else 0